src/extension/alterschema/linter/dangling_additional_items.h

```diff
@@ -0,0 +1,30 @@
+class DanglingAdditionalItems final : public SchemaTransformRule {
+public:
+  DanglingAdditionalItems()
+      : SchemaTransformRule{
+            "dangling_additional_items",
+            "`additionalItems` is ignored when `items` is an object"} {};
+
+  [[nodiscard]] auto
+  condition(const sourcemeta::core::JSON &schema,
+            const sourcemeta::core::JSON &,
+            const sourcemeta::core::Vocabularies &vocabularies,
+            const sourcemeta::core::SchemaFrame &,
+            const sourcemeta::core::SchemaFrame::Location &,
+            const sourcemeta::core::SchemaWalker &,
+            const sourcemeta::core::SchemaResolver &) const
+      -> sourcemeta::core::SchemaTransformRule::Result override {
+    return contains_any(
+               vocabularies,
+               {"https://json-schema.org/draft/2019-09/vocab/applicator",
+                "http://json-schema.org/draft-07/schema#",
+                "http://json-schema.org/draft-06/schema#",
+                "http://json-schema.org/draft-04/schema#"}) &&
+           schema.is_object() && schema.defines("items") &&
+           schema.defines("additionalItems") && schema.at("items").is_object();
```

Comment:
Note that this applies is the schema is a booleans too! I also put comments to clarify this in the rule name and description. I think what you want to check for is if the keyword is set to a schema:

```suggestion
           schema.defines("additionalItems") && is_schema(schema.at("items"));
```